specify/specify7
# Handle incorrect order of setting save blockers

Fixes #4501

Might fix #4508

The problem here is that `setSaveBlockers` overrides previously set blockers. 
There are two saveBlockers which were related to the issues: uniqueness rules and field parsing. 
If the value in the field violated a uniqueness rule and the rules were checked first the save blockers would be set. 
Then when the parse results of the field were being checked and the value matched the correct format, then the business rule save blockers were removed. 

This is not a long term solution, as this bug can easily appear again if a `setSaveBlockers(resource, field, [])` is ran after all saveBlockers for the field are set. 
Furthermore, the current implementation of saveBlockers only allows one error to be shown to a user at a time, if the blockers are set independently. For instance, if the value violates both uniqueness rules and field parsing, only the blocker which is set last will be set and shown to the user. Ideally, all saveBlockers for a given field should be shown to the user at once.

A better solution may be to require passing a `fieldBlockerKey` to uniquely identify each source of a blocker for a field and only modify the blockers with the specific `fieldBlockerKey` with each `setSaveBlockers`. We may be able to implement this already by utilizing the save blocker message. 
Additionally, saveBlockers already support passing a function which has the fetched blockers as a parameter. However, a distinction needs to be made between blockers to correctly set them in the intended way. 

### Checklist

      and self-explanatory (or properly documented)

### Testing instructions
Test #4501 (#4499 is a duplicate of #4501) and #4508 

Additionally, ensure that saveBlockers are being properly set when editing AppResource(s) or Viewset(s). In other words, try and invalidate the file (through syntax error, invalid attribute on a node, etc.)
